File: day1/day1.c
#include <stdio.h>
#include <stdlib.h>

#define ROTOR_SIZE 100
#define ROTOR_INITIAL_POS 50

#define LOGGER_ENABLED 0
#define log(...) if (LOGGER_ENABLED) printf(__VA_ARGS__)
#define log_value(val, fmt) if (LOGGER_ENABLED) printf(" %s = "fmt, #val, (val))
#define print_value(val, fmt) printf("%s = "fmt"\n", #val, (val))

int main(int argc, char *argv[]) {
	if (argc < 2) {
        fprintf(stderr, "Usage: %s <file_input>\n", argv[0]);
		return EXIT_FAILURE;
	}
	FILE *f = fopen(argv[1], "r");
	if (f == NULL) {
		perror("Error opening file");
		return EXIT_FAILURE;
	}

	int rotor = ROTOR_INITIAL_POS;
	size_t zero_clicks = 0;
	size_t zero_stops = 0;

	char line[6];
	while (fgets(line, sizeof(line), f)) {
		int initial_zero = rotor == 0;
		char action = line[0];
		int quant = atoi(&line[1]);
		if (action == 'L') quant *= -1;
		rotor += quant;
		// case change + overflows
		zero_clicks += (rotor <= 0 && !initial_zero) + abs(rotor / ROTOR_SIZE);
		rotor = (rotor % ROTOR_SIZE + ROTOR_SIZE) % ROTOR_SIZE;
		if (rotor == 0) zero_stops++;
		log_value(quant, "%03d");
		log_value(rotor, "%02d");
		log_value(zero_stops, "%04zu");
		log_value(zero_clicks, "%04zu");
		log_value(line, "%s");
	}
	print_value(zero_clicks, "%zu");
	print_value(zero_stops, "%zu");
	fclose(f);
	return EXIT_SUCCESS;
}

File: day1/day1.rs
use std::{
    env,
    fs::File,
    io::{self, BufRead},
};

const ROTOR_SIZE: i32 = 100;
const ROTOR_INITIAL_POS: i32 = 50;

fn main() -> io::Result<()> {
    let args: Vec<String> = env::args().collect();
    let Some(file_path) = args.get(1) else {
        eprintln!(
            "Usage: {} <file_input>",
            args.first().map(|s| s.as_str()).unwrap_or("day1")
        );
        return Err(io::Error::other("file_input missing"));
    };
    let file = File::open(file_path)?;
    let mut rotor = ROTOR_INITIAL_POS;
    let mut zero_clicks = 0;
    let mut zero_stops = 0;
    let reader = io::BufReader::new(file);
    for line_result in reader.lines() {
        let line = line_result?;
        let initial_zero = rotor == 0;
        let action = line.chars().next().unwrap();
        let mut quant: i32 = line[1..].parse().unwrap();
        if action == 'L' {
            quant *= -1;
        }
        rotor += quant;
        zero_clicks += (rotor <= 0 && !initial_zero) as i32 + (rotor / ROTOR_SIZE).abs();
        if rotor == 0 {
            zero_stops += 1;
        }
        rotor = (rotor % ROTOR_SIZE + ROTOR_SIZE) % ROTOR_SIZE;
    }
    println!("zero_clicks = {zero_clicks}");
    println!("zero_stops = {zero_stops}");
    Ok(())
}
